### PR TITLE
Standard material

### PR DESCRIPTION
**Pull request by AdaRoseEdwards**
** Description **

Adding additional material attributes to the Standard material.

Demo: https://samsunginternet.github.io/a-frame-demos/super-standard.html

**Changes proposed:**

Add additional attributes to the Standard material to allow for the use of:

* Spherical Environment Maps
* Displacement Maps
* Normal Maps
* Ambient Occlusion Maps

### CODE REVIEW

**Comment by dmarcos** [comment]
Let's not commit the dist

**Comment by AdaRoseEdwards** [comment]
Sure, I'll revert those out.

File: src/shaders/standard.js [hunk 1]
```diff
@@ -10,29 +10,58 @@ var texturePromises = {};
  */
 module.exports.Component = registerShader('standard', {
   schema: {
+    aoMap: {default: ''},
```
Comment (dmarcos) [inline src/shaders/standard.js]: Can we use a full name instead of an acronym?

File: src/shaders/standard.js [hunk 2]
```diff
@@ -97,10 +138,25 @@ module.exports.Component = registerShader('standard', {
...
 function getMaterialData (data) {
-  return {
+  var newData = {
     color: new THREE.Color(data.color),
     fog: data.fog,
     metalness: data.metalness,
     roughness: data.roughness
   };
+
+  if (data.normalMap) {
```
Comment (dmarcos) [inline src/shaders/standard.js]: make it one liner

File: src/shaders/standard.js [hunk 3]
```diff
@@ -97,10 +138,25 @@ module.exports.Component = registerShader('standard', {
...
+    newData.normalScale = data.normalScale;
+  }
+
+  if (data.aoMap) {
```
Comment (dmarcos) [inline src/shaders/standard.js]: make it one liner

File: src/shaders/standard.js [hunk 4]
```diff
@@ -10,29 +10,58 @@ var texturePromises = {};
...
   },
+
   /**
    * Initializes the shader.
    * Adds a reference from the scene to this entity as the camera.
    */
   init: function (data) {
     this.material = new THREE.MeshStandardMaterial(getMaterialData(data));
     utils.material.updateMap(this, data);
+    if (data.normalMap) utils.material.updateDistortionMap('normal', this, data);
```
Comment (dmarcos) [inline src/shaders/standard.js]: missing brackets after the if

File: src/shaders/standard.js [hunk 5]
```diff
@@ -10,29 +10,58 @@ var texturePromises = {};
...
+    if (data.displacementMap) utils.material.updateDistortionMap('displacement', this, data);
+    if (data.aoMap) utils.material.updateDistortionMap('ao', this, data);
     this.updateEnvMap(data);
   },
 
   update: function (data) {
     this.updateMaterial(data);
     utils.material.updateMap(this, data);
+    if (data.normalMap) utils.material.updateDistortionMap('normal', this, data);
```
Comment (dmarcos) [inline src/shaders/standard.js]: missing brackets after the if

File: src/shaders/standard.js (hunk 1 above)
Comment (AdaRoseEdwards) [inline src/shaders/standard.js]: I can, I was trying to keep the interface the same as THREE.js

Comment (dmarcos) [inline src/shaders/standard.js]: Aframe aimes at being more amicable for non graphics people. Let's use full descriptive names

**Comment by AdaRoseEdwards** [comment]
@dmarcos do you think anything else needs to be done on this PR?

**Comment by ngokevin** [comment]
Would you have time to add some unit tests in `src/tests/shaders/standard.test.js`? Unit tests for setting each kind of map and wrap type? If not, I can help out.

**Comment by AdaRoseEdwards** [comment]
Yeah i'll take a look.

File: src/utils/material.js [hunk 6]
```diff
@@ -28,6 +30,48 @@ module.exports.updateMap = function (shader, data) {
   }
 };
 
+module.exports.updateDistortionMap = function (longType, shader, data) {
```
Comment (ngokevin) [inline src/utils/material.js]: can we add a docstring to this function just describing what it does, its parameters, the longType/shortType mapping?

File: src/utils/material.js [hunk 7]
```diff
@@ -4,6 +4,8 @@
  * @param {object} shader - A-Frame shader instance.
  * @param {object} data
  */
+var THREE = require('../lib/three');
```
Comment (ngokevin) [inline src/utils/material.js]: We should put this at the very top since it's separating a function from its docstring

**Comment by ngokevin** [comment]
Looking good, thanks! If you rebase the PR, the Codecov errors should go away.

File: tests/shaders/standard.test.js [hunk 8]
```diff
@@ -1,7 +1,8 @@
 /* global assert, process, setup, suite, test */
 var entityFactory = require('../helpers').entityFactory;
+var THREE = require('index').THREE;
 
-suite('standard material', function () {
+suite.only('standard material', function () {
```
Comment (ngokevin) [inline tests/shaders/standard.test.js]: can remove this :)

File: src/utils/material.js [hunk 9]
```diff
@@ -29,6 +31,55 @@ module.exports.updateMap = function (shader, data) {
 };
 
 /**
+ * Update's the material's maps which give the illusion of extra geometry.
```
Comment (ngokevin) [inline src/utils/material.js]: typo Update's -> Updates

**Comment by ngokevin** [comment]
Also think it'd be good to put your example scene in `examples/test`

**Comment by ngokevin** [comment]
We can do that now or later ^ your call

**Comment by ngokevin** [comment]
Sorry, one more thing. Is it possible to get coverage for spherical maps after post-load?

screen shot 2016-08-20 at 3 28 15 am

**Comment by AdaRoseEdwards** [comment]
Environment maps don't seem to emit anything once they are loaded I could check in a setInterval until it passes but that is a little hacky. I couldn't see how you were checking loading for cube envMaps.

**Comment by ngokevin** [comment]
Hm, okay. We can do that when we add coverage for envMaps (as well as a bit of refactor).

**Comment by dmarcos** [comment]
We should not make tests optional. I know you want to please people submitting a PR but it's code we will have to maintain.

**Comment by AdaRoseEdwards** [comment]
It's fine I will add a el.emit('envmap-loaded') to both and the write tests for both to listen for that.

**Comment by AdaRoseEdwards** [comment]
I have added tests for environment maps and fixed a bug when loading cube environment maps via 6 urls.

**Comment by AdaRoseEdwards** [comment]
I think that is everything.
I don't feel comfortable resolving the conflicts as my git is showing me it is the whole file in conflict some how?!  Would someone else be able to resolve it?

**Comment by AdaRoseEdwards** [comment]
About the example I will put something together later once I have some assets I feel comfortable sharing. The Ninja is from three.js and sky sphere photo has people in.

**Comment by ngokevin** [comment]
We merged a patch to normalize carriage returns and line feeds. Might be easy to start from master and copy paste your code on. @mkungla might have tips for rebasing

**Comment by dmarcos** [comment]
@AdaRoseEdwards what files do you have conflicts with? Try to replace windows endlines with unix end lines in those files. Commit in your branch and rebase again

**Comment by dmarcos** [comment]
Whe I said commit I meant ammend... @AdaRoseEdwards I'm happy to help if you need

**Comment by ngokevin** [comment]
Trying setting `git config --global core.autocrlf true`

**Comment by AdaRoseEdwards** [comment]
I think that has merged correctly.
I had a weird bit where some code had been duplicated. Fixed. PR looks sane. Tests pass.

**Comment by mkungla** [comment]
@AdaRoseEdwards your last commit went wrong since your merged `master` branch was behind `aframevr/master` 

1. revert this last commit 
```sh
git checkout standard-material
git reset --hard HEAD~1
git push -f origin standard-material
```

2. Make sure you have one `git remote ` pointing to `aframevr/master`
```sh
git remote -v
#should give you output something like this
origin	git@github.com:AdaRoseEdwards/aframe.git (fetch)
origin	git@github.com:AdaRoseEdwards/aframe.git (push)
upstream	git@github.com:aframevr/aframe.git (fetch)
upstream	git@github.com:aframevr/aframe.git (push)
```
if you don't have any remote pointing to `git@github.com:aframevr/aframe.git`
then add it 
```sh
git remote add upstream git@github.com:aframevr/aframe.git
```
then update your forks master, as I see you have also there one conflicting commit "Add more physical maps to stadard material" back this code up and revert this commit as well

```sh
git checkout master
git reset --hard~1
git push -f origin master
```

Now you can rebase your master
```sh
gir fetch --all #do get info from both remotes
git rebase upstream/master
git push -f origin master
```
Now your fork master is in sync with aframe master and you can rebase/merge with your standard-material branch.

**Comment by mkungla** [comment]
#67e1ecd worked and can be squashed when merging but good idea to sort out your fork anyway :)

**Comment by AdaRoseEdwards** [comment]
Thanks for your help, I'll do a rebase on Monday, taking a day off
tomorrow. :)

**Comment by AdaRoseEdwards** [comment]
Apparently I cannot take a break :stuck_out_tongue_closed_eyes: rebased it to make the commits nice.

File: src/utils/material.js [hunk 10]
```diff
@@ -4,6 +4,7 @@
  * @param {object} shader - A-Frame shader instance.
  * @param {object} data
  */
+
```
Comment (ngokevin) [inline src/utils/material.js]: could remove this whitespace line

**Comment by ngokevin** [comment]
@dmarcos lgtm

File: src/shaders/standard.js [hunk 11]
```diff
@@ -57,20 +83,34 @@ module.exports.Component = registerShader('standard', {
...
+    if ((!envMap && !sphericalEnvMap) || this.isLoadingEnvMap) {
       material.envMap = null;
       material.needsUpdate = true;
       return;
     }
     this.isLoadingEnvMap = true;
 
+    // if a spherical env map is defined then use it.
+    if (sphericalEnvMap) {
+      this.el.sceneEl.systems.material.loadTexture(sphericalEnvMap, { src: sphericalEnvMap }, function (texture) {
```
Comment (dmarcos) [inline src/shaders/standard.js]: Let's give a name to the function: `textureLoaded`?

**Pull request merged**